server/api/utils/cache.py

```diff
@@ -0,0 +1,131 @@
+# Copyright 2023 Iguazio
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+import asyncio
+import contextlib
+import typing
+import uuid
+
+
+class CachedObject:
+    """
+    Base class for objects that are cached in the Cache class.
+    """
+
+    def __init__(
+        self,
+        object_id: str,
+        lock: asyncio.Lock = None,
+        expiry_delayed_call: asyncio.Handle = None,
+    ):
+        self._object_id = object_id
+        self.lock = lock or asyncio.Lock()
+        self._expiry_delayed_call: typing.Optional[asyncio.Handle] = expiry_delayed_call
+
+    def matches(self, object_id: str) -> bool:
+        return self._object_id == object_id
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
```

Comment:
To make sure I understand - you're using `__exit__` means that anyone using these cached object is always expected to work in a context manager fashion (i.e. `with...` objects). Right? 
Given that this is an infrastructure that other people may choose to use later, then I suggest that you add some documentation to the class with intended usage and maybe a simple code snip of utilizing it.

Reply:
The idea was that I needed a teardown/cleanup and was inspired by `MLClientCtx` which is not really used as a context manager. I didn't test it yet so not sure if it works. Either way I now realize that I need to ensure it is safe to cancel while someone else might have the lock acquired.